include/aspect/material_model/reaction_model/tian2019_solubility.h

```diff
@@ -0,0 +1,159 @@
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef _aspect_material_reaction_melt_tian2019_solubility_h
+#define _aspect_material_reaction_melt_tian2019_solubility_h
+
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <aspect/melt.h>
+#include <aspect/utilities.h>
+
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    namespace ReactionModel
+    {
+
+      /**
+       * A melt model that calculates melt fraction and entropy change
+       * according to the melting model for dry peridotite of Katz, 2003.
+       * This also includes a computation of the latent heat of melting (if the latent heat
+       * heating model is active).
+       *
+       * These functions can be used in the calculation of melting and melt transport
+       * in the melt_simple material model and can be extended to other material models
+       *
+       * @ingroup ReactionModel
+       */
+      template <int dim>
+      class Tian2019Solubility : public ::aspect::SimulatorAccess<dim>
+      {
+        public:
+          // constructor
+          Tian2019Solubility();
+
+          /**
+           * Declare the parameters this function takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters from the parameter file.
+           */
+          void
+          parse_parameters (ParameterHandler &prm);
+
+
+          /**
+           * Compute the free fluid fraction that can be present in the material based on the
+           * fluid content of the material and the fluid solubility for the given input conditions.
+           * @p in and @p melt_fractions need to have the same size.
+           *
+           * @param in Object that contains the current conditions.
+           * @param melt_fractions Vector of doubles that is filled with the
+           * allowable free fluid fraction for each given input conditions.
+           */
+          double
+          melt_fraction (const MaterialModel::MaterialModelInputs<dim> &in,
+                         const unsigned int porosity_idx,
+                         unsigned int q) const;
+
+          /**
+           * Compute the maximum allowed bound water content at the input
+           * pressure and temperature conditions. This is used to determine
+           * how free water interacts with the solid phase.
+           * @param in Object that contains the current conditions.
+           * @param q unsigned int from 0-3 indexing which rock phase the equilbrium
+           * bound water content is being calculated for
```

Comment:
Oh, is that really the meaning of `q`? We usually use `q` as an index for the quadrature point to index into the vectors that are contained inside `in`. Please clarify and if you use this variable for something else, maybe find a different name unless it has to be `q` for other reasons.